### PR TITLE
Add support for running work-item on epoch end

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Add support for running work-item on epoch end

Part 2 of the fix for issue #217.

Add APIs to create/schedule/delete work items that are tied to an epoch.

Signed-off-by: Alan Jowett <alanjo@microsoft.com>

### CODE REVIEW

File: libs/platform/ebpf_epoch.c [hunk 1]
```diff
@@ -60,12 +60,26 @@ static volatile int64_t _ebpf_current_epoch = 1;
...
+{
+    EBPF_EPOCH_ALLOCATION_MEMORY,
+    EBPF_EPOCH_ALLOCATION_WORK_ITEM,
+} ebpf_epoch_allocation_type_t;
+
 typedef struct _ebpf_epoch_allocation_header
 {
     ebpf_list_entry_t list_entry;
     int64_t freed_epoch;
+    int64_t entry_type;
```
Comment (saxena-anurag) [inline libs/platform/ebpf_epoch.c:73]: ebpf_epoch_allocation_type_t entry_type?

File: libs/platform/ebpf_epoch.c [hunk 2]
```diff
@@ -372,3 +397,40 @@ _ebpf_flush_worker(void* context)
...
+
+    work_item->callback = callback;
+    work_item->callback_context = callback_context;
+    work_item->header.entry_type = EBPF_EPOCH_ALLOCATION_WORK_ITEM;
+
+    return work_item;
+}
+
+void
+ebpf_epoch_schedule_work_item(ebpf_epoch_work_item_t* work_item)
```
Comment (shankarseal) [inline libs/platform/ebpf_epoch.c:417]: When and why will a WI be scheduled on epoch end?

Comment (Alan-Jowett) [reply libs/platform/ebpf_epoch.c:417]: Code needs to delete the pages allocated to hold code. We can't use the normal memory allocator as it isn't page aligned.

File: libs/platform/ebpf_epoch.c (hunk 1 above)
Comment (dthaler) [reply libs/platform/ebpf_epoch.c:73]: +1

File: libs/platform/ebpf_epoch.c (hunk 2 above)
Comment (dthaler) [reply libs/platform/ebpf_epoch.c:417]: Please add a // comment somewhere relevant explaining that.

**Review by dthaler — APPROVED** [review]

**Review by poornagmsft — APPROVED** [review]

**Pull request merged**